lib/collection/src/shards/local_shard.rs

```diff
@@ -395,9 +395,46 @@ impl LocalShard {
         bar.set_style(progress_style);
 
         bar.set_message(format!("Recovering collection {collection_id}"));
+
+        // To optimize loading we want to start reading (and applying) WAL from the last operation
+        // persisted by the storage.
+        //
+        // Each `Segment` is versioned individually, so there's no *single* storage version.
+        // We can guarantee that the *minimal* (i.e., "oldest") persisted version among *all*
+        // `Segment`s has been successfully applied to *all* of them.
+        //
+        // There's no public interface to access `Segment::persisted_version`, but
+        // `LocalShard::load` loads segments state from disk and then calls
+        // `LocalShard::load_from_wal`. So, as each `Segment` was *just* loaded from disk
+        // and was not modified in any way yet, `Segment::version` (returned by
+        // `<Segment as SegmentEntry>::version`) and `Segment::persisted_version` are the same.
+
         let segments = self.segments();
-        // ToDo: Start from minimal applied version
-        for (op_num, update) in wal.read_all() {
+
+        let min_persisted_version = segments
+            .read()
+            .iter()
+            .map(|(_, segment)| segment.get().read().version())
+            .min()
+            .unwrap_or(0);
+
+        // `0` is a valid version.
+        //
+        // `Segment::version` declared as `Option<SeqNumberType>`.
+        //
+        // However, `<Segment as SegmentEntry>::version` returns `self.version.unwrap_or(0)`,
+        // which means we can't precisely distinguish an empty `Segment` (i.e., no operations applied)
+        // from a `Segment` with version `0` (i.e., exactly one operation with version `0` applied).
+        //
+        // To work around this corner case, we explicitly check that `min_persisted_version > 0`,
+        // so that if `min_persisted_version == 0` we won't skip operation `0` and will always
+        // (re)apply it.
+
+        let ops = wal.read_all().skip_while(|&(op_num, _)| {
+            op_num <= min_persisted_version && min_persisted_version > 0
+        });
```

Comment:
Alternatively, we can simply use `op_num < min_persisted_version`, but then we will always reapply the last persisted operation. (But I'd leave the explanation comment in any case, if we ever fix some obscure bug regarding "WAL entry `0`".)